Allow custom selector types to be defined

Mathe172 opened the pull request:
This PR is an implementation of the proposed changes discussed in #3451 (as a new PR since it works completely different). See #3353 for an overview of what it attempts to achieve.

The goal is to provide a way to allow for more than just adding new parameters to filter entities - this is done by allowing new prefixes to be registered. Each time a selector is parsed, the best matching prefix (i.e. the longest matching one) is selected and the associated `SelectorHandler` is called.

(Note: There may be a better/faster way to do the lookup, e.g. using Tries or similar, but this would require more code)

Commit:
Allow custom selector types to be defined

Rerouted EntitySelector.matchEntities(ICommandSender, String, Class), .isSelector(String) and .matchesMultiplePlayers(String) over SelectorHandlerManager
Added SelectorHandlerManger to handle calls to EntitySelector methods and choose appropriate SelectorHandler
Added SelectorHandler
The vanilla version of SelectorHandler uses the old implementations, which are renamed to *Default(...) and is registered for prefixes @p,@a,@r,@e and as fallback
Adjusted forge.exc to correctly obfuscate parameters of the new *Default(...) methods
Added test mod to demonstrate the capabilities
Test mod registers the "self"-selector, which returns the entity executing the command (as @s and @es (to provoke warning))

liach commented on src/main/java/net/minecraftforge/common/command/SelectorHandler.java:
```diff
@@ -0,0 +1,52 @@
+ */
+public abstract class SelectorHandler
+{
+    public abstract <T extends Entity> List<T> matchEntities(ICommandSender sender, String token, Class<? extends T> targetClass) throws CommandException;
```
Javadoc please

liach commented on src/main/java/net/minecraftforge/common/command/SelectorHandlerManager.java:
```diff
@@ -0,0 +1,158 @@
+
+    //the ordering is reversed such that longer prefixes appear before their shorter substrings
+    private static final SortedMap<String, SelectorHandler> selectorHandlers = new TreeMap<String, SelectorHandler>(Collections.<String> reverseOrder());
+    private static final SortedMap<String, String> registeringMods = new TreeMap<String, String>(Collections.<String> reverseOrder());
```
I wonder if there is some sort of trie in the libraries.

PhiProven replied:
We couldnt't find one.
There is a `Trie` in the `com.ibm.icu` lib, but it's not really handy.
And there's an implementation in `org.apache.commons-collections4`, but that lib is not included in the minecraft dependencies and the implementation does not allow for incremental search (i.e. passing one character after another), so that wouldn't really make things nicer or faster.

We could just stack `Character` maps in order to achieve the incremental search, but it makes the code a bit unreadable, so we didn't do this.
Should we add this for performance reasons? (Though it shouldn't really matter as long as there are not too many prefixes and they are not too long)

liach replied:
Only add one written instead of another stupid library. Then just left it like that is good.

liach reviewed (commented):
Things look good!

liach commented on src/main/java/net/minecraftforge/common/command/SelectorHandlerManager.java:
```diff
@@ -0,0 +1,158 @@
+        for (final String prefix : ArrayUtils.toArray("@p", "@a", "@r", "@e"))
+        {
+            selectorHandlers.put(prefix, vanillaHandler);
+            registeringMods.put(prefix, "Minecraft");
```
Should `Minecraft` be lowercase?

kashike replied:
Mod ids should be lowercase now, so I'd say yes.

bs2609 commented on src/main/java/net/minecraftforge/common/command/SelectorHandlerManager.java:
```diff
@@ -0,0 +1,158 @@
+     */
+    public static boolean isShadowed(final String prefix)
+    {
+        for (final String other : registeringMods.keySet())
```
Would this work (using `NavigableMap`)?
`String other = registeringMods.floorKey(prefix);`
`return prefix.startsWith(other);`

Mathe172 replied:
Unfortunately, I don't think so:
Assume you have prefixes `a` and `aa` registered. When you call this method with `ac`, the current code would return `true` (since `"ac".startsWith("a")`). `floorKey` however returns `null` (when the order is reversed) and `ceilingKey` returns `aa`, which is also not what we want.

`headMap`/`tailMap` (as you suggested for `getHandler`) or maybe even `subMap` should work however - we'll look into it and update the PR accordingly.

bs2609 commented on src/main/java/net/minecraftforge/common/command/SelectorHandlerManager.java:
```diff
@@ -0,0 +1,158 @@
+     */
+    public static boolean isShadowing(final String prefix)
+    {
+        for (final String other : registeringMods.keySet())
```
Similarly,
`String other = registeringMods.ceilingKey(prefix);`
`return other.startsWith(prefix);`

Mathe172 replied:
See my comment for `isShadowed`

bs2609 commented on src/main/java/net/minecraftforge/common/command/SelectorHandlerManager.java:
```diff
@@ -0,0 +1,158 @@
+     */
+    public static SelectorHandler getHandler(final String selectorStr)
+    {
+        for (final Entry<String, SelectorHandler> handler : selectorHandlers.entrySet())
```
This could also probably be optimised.
At least by using `selectorHandlers.headMap(selectorStr, true).entrySet()` to trim the range to iterate over.

Mathe172 replied:
This should definitely work - it may even be possible to use `selectorHandlers.subMap(selectorStr, true, selectorStr.substring(0,1), true).entrySet()`

bs2609 reviewed (commented):
Seems fairly clean.
It looks like this could be improved by using the extra methods provided by `NavigableMap` to cut down on the amount of iterating required.

Commit:
Updated SelectorHandlerManager and SelectorHandler

Added doc to SelectorHandler.matchEntities
Changed SelectorHandlerManager.isShadowed and .isShadowing to .getShadowing and .getShadowed, which now return a collection of colliding prefixes
Mod id of vanilla Minecraft is now lowercase
Optimized performance by using NavigableMap.subMap and .headMap

Mathe172 commented:
Updated the PR based on the suggestions from @bs2609 and @liach. Also changed the methods `isShadowed` and `isShadowing` to `get...` versions that return collections of all colliding prefixes. This may enable a mod to react on collisions more accurately.

Commit:
Changed SelectorHandlerManager to use modId instead of mod name

bs2609 commented on src/main/java/net/minecraftforge/common/command/SelectorHandlerManager.java:
```diff
@@ -0,0 +1,170 @@
+    {
+        Map<String, String> ret = new HashMap<String, String>();
+
+        for (final Entry<String, String> other : registeringMods.headMap(prefix).entrySet())
```
This should be `headMap(prefix, true)`

Mathe172 replied:
The idea was that this edge case is already handled by `getShadowed()`. Changing this would possibly emit two warnings.

bs2609 replied:
I think `getShadowed` and `getShadowing` need a note in the Javadoc specifing their behaviour regarding exact matches in this case.

bs2609 commented on src/main/java/net/minecraftforge/common/command/SelectorHandlerManager.java:
```diff
@@ -0,0 +1,170 @@
+        for (final Entry<String, String> other : registeringMods.headMap(prefix).entrySet())
+            if (other.getKey().startsWith(prefix))
+                ret.put(other.getKey(), other.getValue());
+
```
Needs checking, but I think you can return as soon as the condition fails

bs2609 replied:
The reverse ordering is confusing me more than it probably should >_>
This (previous comment) requires a `descendingMap()` call between `headMap` and `entrySet` in the loop declaration.

Mathe172 replied:
The intention was to return all colliding prefixes, not just one.

Calling `descendingMap()` is not needed: Either you return all (in which case it doesn't matter), or you return one - this should probably be the longest one, which is the first in the original map (i.e. without `descendingMap()`)

bs2609 replied:
To clarify, it's my suggestion that required the use of a `descendingMap` call.

The point is, if you want to find all registered prefixes starting with "a", it makes sense to iterate through the map starting at "a". As it stands, using `headMap`, "a" would be the last element visited. Hence the `descendingMap` call.

Then, when looking for prefixes starting with "a", you can stop when you get to "b", because no subsequent prefixes will satisfy the condition.

Doing so should still return all colliding prefixes, the aim is simply to avoid unnecessarily checking elements.

bs2609 reviewed (commented):
Looks improved! A couple of small pointers, and personally I'd put braces on if/for statement bodies if they are on a new line, but I don't know what the official Forge position would be for that.

Commit:
Updated SelectorHandlerManager.getShadowing

getShadowing now short circuits when first non-matching prefix found (requires map to be iterated in reverse order)

Mathe172 commented:
Updated according to suggestions from @bs2609

liach commented on src/main/java/net/minecraftforge/common/command/SelectorHandlerManager.java:
```diff
@@ -0,0 +1,171 @@
+    public static SelectorHandler getHandler(final String selectorStr)
+    {
+        if (!selectorStr.isEmpty())
+            for (final Entry<String, SelectorHandler> handler : selectorHandlers.subMap(selectorStr, true, selectorStr.substring(0, 1), true).entrySet())
```
Why `selectorStr.substring(0, 1)`

bs2609 replied:
That's the last string in the map it could possibly start with.

liach replied:
Why not just `selectorHandlers.subMap(selectorStr, true, selectorStr, true).entrySet()`, it should function the same

bs2609 replied:
That would just return at most single entry, not a range.

Commit:
Reworded docs and avoided IndexOutOfBoundsException

Changed docs of getShadowed to be more precise
Prevented IndexOutOfBoundsException in getShadowed

Commit:
Updated formatting to match Forge standards

bs2609 commented on src/main/java/net/minecraftforge/common/command/SelectorHandlerManager.java:
```diff
@@ -0,0 +1,199 @@
+     *
+     * @return A {@link Map} mapping the prefixes to the mod ids registering them
+     */
+    public static Map<String, String> getShadowing(final String prefix)
```
You could add a check for empty prefixes here too, though it's not required for correctness.
`if (prefix.isEmpty()) return new HashMap<String, String>(registeringMods);`

liach replied:
```java
if (prefix.isEmpty())
    return Maps.newHashMap(registeringMods);
```

Mathe172 replied:
You should never have to call this method with an empty string (except if you wanted to get a list of all registered prefixes for some reason, but in this case I'd rather provide a special method), since empty prefixes are not allowed to be registered.

The only reason for the check in `getShadowed` is to prevent the `IndexOutOfBoundsException` which would otherwise occur

Commit:
Updated documentation of SelectorHandler.isSelector

SelectorHandler.isSelector doc now recommends to throw exceptions in matchEntities instead of returning false, as this alone is not sufficient to prevent matchEntities from being called
Updated test mod to check syntax of selector and throw if it is wrong

mezz commented on src/main/java/net/minecraftforge/common/command/SelectorHandler.java:
```diff
@@ -0,0 +1,58 @@
+     * (Since {@link net.minecraft.command.EntitySelector#matchEntities(ICommandSender, String, Class) EntitySelector.matchEntities}
+     * returns an empty list when the string is illformed instead of throwing an exception (as is indirectly done when this method returns {@code false})
+     */
+    public boolean isSelector(String selectorStr)
```
This javadoc is extremely confusing, can you clarify it at all and make it wrap to shorter line lengths?

Mathe172 replied:
To be fair, the usage of `isSelector` is itself very confusing ;)
This is what we know about the vanilla usage of the method (maybe you have an idea on how to write that down, as we've already tried for more than an hour):
There are a few places where the following logic is more or less implemented (this example is from `CommandBase.getEntityList`):
* If `isSelector` returns `true`, `matchEntities` is called (which throws if a syntax error is encountered)
* If not, the string is passed to `getEntity`, which calls `matchOneEntity`
  * There, `matchEntities` is called again, which obviously fails (by returning an empty list, and *not* by throwing)
  * Then, the string is interpreted as player name, UUID, ...
  * If everything fails, an exception is thrown

We need to override it in order to ensure that `matchEntities` is called for custom prefixes.

As you can see, there are two possible reactions when the syntax is invalid:
* Throw an exception. This is easy to implement and works most of the time (that's why the doc suggests it). This has two possible downsides:
  * It breaks when the string could be a player name/UUID (for reasonable prefixes, this shouldn't be the case)
  * In case of the `/say` command, an error is shown instead of the unparsed selector (this happens in vanilla for any syntax error that isn't caught by the regex)
* Return `false` in `isSelector` *and* an empty list in `matchEntitites`. This however causes the whole mechanism (matching the prefix and checking the syntax) to be called twice (unnecessarily) and requires more care from the modder.

We thought it's easiest to default to `true` and change the implementation for the vanilla case. This is what the doc tries to explain briefly (as the method would look very strange otherwise).

We see three options:
* Improve the doc (we would be very happy about any suggestions ;) )
* Require vanilla behaviour from the modder (i.e. check whether the string seems reasonable (whatever that means for that specific selector) and return an empty list)
* Return `true` whenever a prefix is encountered and patch vanilla to work with that (this behaves differently for the `/say` command in a few cases). This would allow us to remove the method entirely from `SelectorHandler`

What would you suggest?
(sorry for the long answer, we again failed to explain it in a simpler way ;) )

mezz commented on src/main/java/net/minecraftforge/common/command/SelectorHandler.java:
```diff
@@ -0,0 +1,58 @@
+/**
+ * Handler for custom types of selectors registered with {@link SelectorHandlerManager}
+ */
+public abstract class SelectorHandler
```
Is there any reason this is an abstract class instead of an interface?

Mathe172 replied:
It's needed for the default implementation of `isSelector` (assuming you want backwards compatibility to Java 7 and lower). This way, you shouldn't have to worry about this method too much as long as you don't do anything strange.

Commit:
Changed behavior of SelectorHandler.isSelector

Changed behavior of SelectorHandler.isSelector and .matchEntities to conform with Vanilla
Removed default implementation of SelectorHandler.isSelector and changed SelectorHandler from abstract class to interface
Updated SelectorHandlerTest to the new behavior

 Fixed possible NPE in SelectorHandlerTest

PhiProven commented:
Changed the behavior of `SelectorHandler.isSelector()`.
Docs should be simpler now. Should we use any of the other options mentioned above though, as they are easier to use for the modder?

mezz commented:
I just finished reading and understanding the previous behavior, what you have now seems much more clear.

I think if you want to add more functionality/power to `isSelector`, you should rename it and break away from vanilla entirely and have it return an enum based on the behavior you want. Then have your `SelectorHandlerManager` do whatever is necessary to translate that into the vanilla madness.

Mathe172 commented:
I don't think we need anything more powerful than `isSelector` - you can already do any checks you might want in either `isSelector` or `matchEntities`. If you think the current doc is reasonable, I would just leave it as it is.

mezz commented:
Ok looks good then. Tested and the test selectors work as expected.

Why is there an info line printed when a new selector shadows another?
It seems like something that is handled normally and not worth logging about, unless I misunderstand it.

Commit:
Removed shadowing logs from SelectorHandlerManager

Updated SelectorHandlerTest as the shadowing test is no longer needed

PhiProven commented:
Ok, removed it.

The idea was to print a little info for the user, in case the behavior of some selectors change because of another handler shadowing it. But it's probably not needed, as the user should be aware of which mods add which prefixes.

PhiProven commented:
Should we tweak the prefix-resolution system by the following?
Instead of just using the longest matching prefix, we could use the longest one for which `isSelector` actually returns `true`.
This way, shadowing by other handlers shouldn't cause unexpected side effects.
Downside of this is, that the prefix-resolution would be more expensive (and has to be done twice, once for `isSelector` and once for `matchEntities`).

This could also simplify the doc for `isSelector` and `matchEntities`: As `matchEntities` of a handlers is only called if `isSelector` returns `true` (except for the Vanilla one, which is used if no other prefix applies), `matchEntities` is no longer required to return an empty list, if `isSelector` returns `false`.

PhiProven commented:
After a second thought, it's probably best, if we leave the prefix-resolution system as it is. Any mod that registers a custom prefix can use `getShadowing` and `getShadowed` and use that information to take care of such shadowing conflicts itself.

mezz commented:
I'm not sure performance is a big issue for that particular case.
Also, can you give a concrete example where this shadowing control is used?

PhiProven commented:
As explained in the previous versions of this PR, we want to create a completely new framework for command and selector parsing by introducing a typesystem, etc. We want to register a `SelectorHandler` for the prefix `#@` (or something like that) for the selectors registered through our mod. Only registering this one prefix for our handler, and not one for every selector which is registered with our mod, is more flexible. It allows us to change selectors at server restart or have different selectors for client, server, etc.
If (for whatever reason) some other mod registers a selector `#@a`, then a selector `@ab` which is registered through our mod would not work properly. As it had to be called with `#@ab`, but we only registered the prefix `#@` the call wouldn't be delegated to us but the handler for `#@a`.
We can resolve this by just registering our prefix `#@` as late as possible and take care of this conflict.
As this shouldn't occur for simple selectors (instead of whole domains), I think it's best if we take care of that in our mod ourselves.
Hope, the explanation wasn't to confusing :)

mezz commented:
I'm jumping between a lot of open PRs, I apologize if I'm asking you to repeat yourself.

> It allows us to change selectors at server restart or have different selectors for client, server, etc.

Why do you want that?

PhiProven commented:
That probably was a bad example, sorry for that.
My point was, that there theoretically could be situations when you have to take a little care of such shadowing conflicts.
But the current state of the PR already allows for that via `getShadowed` and `getShadowing`. As this shouldn't be a problem for simple selectors, I think it's best to leave the PR as it is now.

mezz commented:
One of the requirements for PRs is that they do only what is necessary to accomplish their goal.

If you do not have a concrete use for prefix-based selectors, then I think it would make the code significantly simpler to remove that functionality. Simplicity helps other modders make good use of it, and helps ensure it's ported correctly to new Minecraft versions.

I agree that it is working correctly, but the added complexity does not seem justified based on what you've said so far. If I understand correctly, without prefix resolution and shadowing you would still be able to use "domains" as a convention, the only difference is that you would have to register all your commands.

PhiProven commented:
Requiring the whole selector name instead of the current prefix based mechanism wouldn't simplify anything for the modder (as you don't have to worry about that in case you just add a single selector).
The prefix mechanism just allows for a litlle more flexibility.

Mathe172 commented:
Any update on this?
If needed, we could for example simplify the code by removing the `getShadowed` and `getShadowing` methods in favor of just exposing `selectorHandlers` and `registeringMods`. This would reduce the complexity of the PR without really sacrificing any functionality from a modders perspecitve.

PhiProven commented:
@mezz Any updates? (See comment above)

Commit:
Removed shadowing related methods

PhiProven commented:
Removed shadowing related methods.
Rebased to resolve merge conflicts.

liach commented:
This is a useful feature. please do not kill it.

liach commented:
@mathe172 can you update the licenses with a gradle task?

liach commented:
Ah this one should be closed. Sorry for my error because the pull for 1.12 is already merged.
